### DISCUSSION TITLE
Generating RDF collections

### DISCUSSION THREAD

**Discussion by mielvds** [SPARQL-Anything/sparql.anything#214] [Q&A]
Hi! I was wondering whether SPARQL anything can support the construction of an [RDF Collection](https://www.w3.org/TR/rdf-schema/#h3_ch_collectionvocab), eg. in cases where order of a CSV or JSON array needs to be mapped to as `rdf:List` (like for the object of `skos:memberList`).

1. Do the functions with container membership properties extend to collections as well, or can they be made that way?
2. Creating a collection is difficult in SPARQL 1.1. Is it possible to add a function that creates a collection, like `fx:list(?a, ?l)`, which adds `?a` to a list `?l`?

**Comment by enridaga** [comment]
Thank you, this is an interesting question! 

> Do the functions with container membership properties extend to collections as well, or can they be made that way?

They only work with container membership properties, querying `rdf:List` is quite hard in SPARQL (and you don't need to to query Facade-X as collections are represented as containers, like `rdf:Seq` and not `rdf:List`). There is no plan at the moment of adding functions for `rdf:List`. However, you can find something in Apache Jena functions that should work with SPARQL Anything as well: https://jena.apache.org/documentation/query/library-propfunc.html

Personally, I dislike `rdf:List` as I find it quite hard to work with in comparison to container membership properties. FYI, the query-ability of the different list representations in RDF was analysed in detail in this article: https://content.iospress.com/articles/semantic-web/sw210436 

> Creating a collection is difficult in SPARQL 1.1. Is it possible to add a function that creates a collection, like fx:list(?a, ?l), which adds ?a to a list ?l?

I am not sure how such a function could work in SPARQL (but see Jena functions for some functions with `rdf:List`).

However, **you don't need any special feature to build RDF Collections with SPARQL**, see the following example.

Input data (`data.csv`):
```csv
item-1
item-2
item-3
item-4
item-4
item-6
```

Query `make-list.sparql`:

```sparql
PREFIX owl:  <http://www.w3.org/2002/07/owl#>
PREFIX rdf:  <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
PREFIX fx:   <http://sparql.xyz/facade-x/ns/>
PREFIX xyz:  <http://sparql.xyz/facade-x/data/>

CONSTRUCT {
 ?item a rdf:List .
 ?item rdf:first ?value .
 ?item rdf:rest ?rest .
} WHERE {
#SELECT * WHERE {
	SERVICE <x-sparql-anything:> {
		fx:properties fx:location "./data.csv" .
		?csv a fx:root .
		?csv ?li ?item .
		?item rdf:_1 ?value
		OPTIONAL { ?csv ?li2 ?next . filter (fx:next(?li) = ?li2) }
		BIND (IF(BOUND(?next), ?next, rdf:nil) AS ?rest )
	}
}
```

Output of `$ java -jar sparql-anything-0.6.0.jar -q make-list.sparql`

```turtle
@prefix fx:   <http://sparql.xyz/facade-x/ns/> .
@prefix owl:  <http://www.w3.org/2002/07/owl#> .
@prefix rdf:  <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix xyz:  <http://sparql.xyz/facade-x/data/> .

[ rdf:type   rdf:List ;
  rdf:first  "item-1" ;
  rdf:rest   [ rdf:type   rdf:List ;
               rdf:first  "item-2" ;
               rdf:rest   [ rdf:type   rdf:List ;
                            rdf:first  "item-3" ;
                            rdf:rest   [ rdf:type   rdf:List ;
                                         rdf:first  "item-4" ;
                                         rdf:rest   [ rdf:type   rdf:List ;
                                                      rdf:first  "item-4" ;
                                                      rdf:rest   [ rdf:type   rdf:List ;
                                                                   rdf:first  "item-6" ;
                                                                   rdf:rest   ()
                                                                 ]
                                                    ]
                                       ]
                          ]
             ]
] .

```

**Comment by mielvds** [comment]
Thanks @enridaga, that query is quite neat and solves the problem! In fact, I think a repository of query templates for common and not so common patterns would be a nice addition of the website or docs. Patterns like that "could" also be captured in functions such as the jena ones for usability sake.

> Personally, I dislike `rdf:List` as I find it quite hard to work with in comparison to container membership properties.

Me neither, but if SKOS tells you it's `rdf:List`, it's `rdf:List` :)

**Comment by enridaga** [comment]
> I think a repository of query templates for common and not so common patterns would be a nice addition of the website or docs.

Indeed :)
